File: app/src/main/java/com/ruideraj/secretelephant/injection/AppModule.java
package com.ruideraj.secretelephant.injection;

import android.app.Application;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {
    private Application mApplication;

    public AppModule(Application application) {
        mApplication = application;
    }

    @Provides
    public Application providesApplication() {
        return mApplication;
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/injection/AppModule.kt
package com.ruideraj.secretelephant.injection

import android.app.Application
import android.content.Context
import com.ruideraj.secretelephant.PermissionManager
import com.ruideraj.secretelephant.PermissionManagerImpl
import com.ruideraj.secretelephant.Preferences
import com.ruideraj.secretelephant.PreferencesImpl
import dagger.Module
import dagger.Provides

@Module
class AppModule constructor (private val application: Application) {

    @Provides
    fun providesApplication() = application

    @Provides
    fun providesContext(): Context = application

    @Provides
    fun providesPermissionManager(permissionManagerImpl: PermissionManagerImpl): PermissionManager
            = permissionManagerImpl

    @Provides
    fun providesPreferences(preferencesImpl: PreferencesImpl): Preferences = preferencesImpl
}